bluepysnap/query.py

```diff
@@ -0,0 +1,116 @@
+from collections.abc import Mapping
+from copy import deepcopy
+
+import numpy as np
+import pandas as pd
+from bluepysnap.exceptions import BluepySnapError
+from bluepysnap import utils
+
+# this constant is not part of the sonata standard
+NODE_ID_KEY = "node_id"
+EDGE_ID_KEY = "edge_id"
+POPULATION_KEY = "population"
+OR_KEY = "$or"
+AND_KEY = "$and"
+NODE_SET_KEY = "$node_set"  # the only key that is not a part of QUERY_KEYS
+QUERY_KEYS = {NODE_ID_KEY, EDGE_ID_KEY, POPULATION_KEY, OR_KEY, AND_KEY}
+
+
+def _positional_mask(data, node_ids):
+    """Positional mask for the node IDs.
+
+    Args:
+        node_ids (None/numpy.ndarray): the ids array. If None all ids are selected.
+
+    Examples:
+        if the data set contains 5 nodes:
+        _positional_mask([0,2]) --> [True, False, True, False, False]
+    """
+    if node_ids is None:
+        return np.full(len(data), fill_value=True)
+    mask = np.full(len(data), fill_value=False)
+    valid_node_ids = pd.Index(utils.ensure_list(node_ids)).intersection(data.index)
+    mask[valid_node_ids] = True
+    return mask
+
+
+def _circuit_mask(data, population_name, queries):
+    """Handle the population, node ID queries."""
+    populations = queries.pop(POPULATION_KEY, None)
+    if populations is not None and population_name not in set(utils.ensure_list(populations)):
+        node_ids = []
+    else:
+        node_ids = queries.pop(NODE_ID_KEY, None)
+    return queries, _positional_mask(data, node_ids)
+
+
+def _properties_mask(data, population_name, queries):
+    """Return mask of IDs matching `props` dict."""
+    # pylint: disable=assignment-from-no-return
+    unknown_props = set(queries) - set(data.columns) - QUERY_KEYS
+    if unknown_props:
+        return np.full(len(data), fill_value=False)
+
+    queries, mask = _circuit_mask(data, population_name, queries)
+    if not mask.any():
+        # Avoid fail and/or processing time if wrong population or no nodes
+        return mask
+
+    for prop, values in queries.items():
+        prop = data[prop]
+        if np.issubdtype(prop.dtype.type, np.floating):
+            v1, v2 = values
+            prop_mask = np.logical_and(prop >= v1, prop <= v2)
+        elif isinstance(values, str) and values.startswith('regex:'):
+            prop_mask = np.logical_and(np.full(len(prop), True), prop.str.match(values[6:] + "\\Z"))
+        else:
+            prop_mask = np.in1d(prop, values)
+        mask = np.logical_and(mask, prop_mask)
+    return mask
+
+
+def traverse_queries_bottom_up(queries, node_function):
```

Comment:
The key change is this traversing function

Reply:
I think it clearer than the current implementation

Reply:
I think the name is not perfect. 
What about : 
transform_queries_bottom_up(queries, transform_function)

Because the idea is to transform the queries using a transform_function from the bottom to up. 
Also this makes sense only if the current "node_function" is a inplace transform function of the queries if I understand correctly.

Reply:
Idea is not transform but to traverse. Transforming is collateral here due to resolving of `$node_set`.

Reply:
Here is the function that does explicit transform + it does not require deep copying. From functions point of view it is more proper classic variant of the function. One problem with it. It is hard to read/understand:
```
def transform_queries_bottom_up(queries, transform):
    """Traverse queries tree from leaves to root, left to right.

    Args:
        queries (dict): queries
        transform (function): function to execute on each node of `queries` in traverse order
    """
    transformed_queries = {}
    for key in list(queries.keys()):
        if key in {OR_KEY, AND_KEY}:
            transformed_queries[key] = []
            for subquery in queries[key]:
                transformed_queries[key].append(transform_queries_bottom_up(subquery, transform))
            # `transform` has to be called on `transformed_queries`
            transformed_queries[key] = transform(transformed_queries, key)
        elif isinstance(queries[key], Mapping):
            if VALUE_KEYS & set(queries[key]):
                if not set(queries[key]).issubset(VALUE_KEYS):
                    raise BluepySnapError("Value operators can't be used with plain values")
            else:
                transformed_queries[key] = transform_queries_bottom_up(queries[key], transform)
                # `transform` has to be called on `transformed_queries`
                transformed_queries[key] = transform(transformed_queries, key)
        else:
            transformed_queries[key] = transform(queries, key)
    return transformed_queries
```

Reply:
I see your point. 
Gosh, it appears to be more complex design wise than first expected (at least to me)!
Ok lets keep it like this. Maybe having the edges ids in the future will change the function again for some reasons. So I think your impl is good at the moment.